File: packages/node/iii/src/bridge-types.ts
export enum MessageType {
  RegisterFunction = 'registerfunction',
  RegisterService = 'registerservice',
  InvokeFunction = 'invokefunction',
  InvocationResult = 'invocationresult',
  RegisterTriggerType = 'registertriggertype',
  RegisterTrigger = 'registertrigger',
  UnregisterTrigger = 'unregistertrigger',
  UnregisterTriggerType = 'unregistertriggertype',
  TriggerRegistrationResult = 'triggerregistrationresult',
  WorkerRegistered = 'workerregistered',
}

export type RegisterTriggerTypeMessage = {
  type: MessageType.RegisterTriggerType
  id: string
  description: string
}

export type UnregisterTriggerTypeMessage = {
  type: MessageType.UnregisterTriggerType
  id: string
}

export type UnregisterTriggerMessage = {
  type: MessageType.UnregisterTrigger
  id: string
}

export type TriggerRegistrationResultMessage = {
  type: MessageType.TriggerRegistrationResult
  id: string
  trigger_type: string
  function_path: string
  result?: any
  error?: any
}

export type RegisterTriggerMessage = {
  type: MessageType.RegisterTrigger

  id: string
  /**
   * The type of trigger. Can be 'cron', 'event', 'http', etc.
   */
  trigger_type: string
  /**
   * Engine path for the function, including the service and function name
   * Example: software.engineering.code.rust
   * Where software, engineering, and code are the service ids
   */
  function_path: string
  config: any
}

export type RegisterServiceMessage = {
  type: MessageType.RegisterService
  id: string
  description?: string
  parent_service_id?: string
}

export type RegisterFunctionFormat = {
  name: string
  /**
   * The description of the parameter
   */
  description?: string
  /**
   * The type of the parameter
   */
  type: 'string' | 'number' | 'boolean' | 'object' | 'array' | 'null' | 'map'
  /**
   * The body of the parameter
   */
  body?: RegisterFunctionFormat[]
  /**
   * The items of the parameter
   */
  items?: RegisterFunctionFormat
  /**
   * Whether the parameter is required
   */
  required?: boolean
}

export type RegisterFunctionMessage = {
  type: MessageType.RegisterFunction
  /**
   * The path of the function
   */
  function_path: string
  /**
   * The description of the function
   */
  description?: string
  /**
   * The request format of the function
   */
  request_format?: RegisterFunctionFormat
  /**
   * The response format of the function
   */
  response_format?: RegisterFunctionFormat
  metadata?: Record<string, unknown>
}

export type InvokeFunctionMessage = {
  type: MessageType.InvokeFunction
  /**
   * This is optional for async invocations
   */
  invocation_id?: string
  /**
   * The path of the function
   */
  function_path: string
  /**
   * The data to pass to the function
   */
  data: any
  /**
   * W3C trace-context traceparent header for distributed tracing
   */
  traceparent?: string
  /**
   * W3C baggage header for cross-cutting context propagation
   */
  baggage?: string
}

export type InvocationResultMessage = {
  type: MessageType.InvocationResult
  /**
   * The id of the invocation
   */
  invocation_id: string
  /**
   * The path of the function
   */
  function_path: string
  result?: any
  error?: any
  /**
   * W3C trace-context traceparent header for distributed tracing
   */
  traceparent?: string
  /**
   * W3C baggage header for cross-cutting context propagation
   */
  baggage?: string
}

export type FunctionInfo = {
  function_path: string
  description?: string
  request_format?: RegisterFunctionFormat
  response_format?: RegisterFunctionFormat
  metadata?: Record<string, unknown>
}

export type WorkerStatus = 'connected' | 'available' | 'busy' | 'disconnected'

export type WorkerInfo = {
  id: string
  name?: string
  runtime?: string
  version?: string
  os?: string
  ip_address?: string
  status: WorkerStatus
  connected_at_ms: number
  function_count: number
  functions: string[]
  active_invocations: number
}

export type WorkerRegisteredMessage = {
  type: MessageType.WorkerRegistered
  worker_id: string
}

export type BridgeMessage =
  | RegisterFunctionMessage
  | InvokeFunctionMessage
  | InvocationResultMessage
  | RegisterServiceMessage
  | RegisterTriggerMessage
  | RegisterTriggerTypeMessage
  | UnregisterTriggerMessage
  | UnregisterTriggerTypeMessage
  | TriggerRegistrationResultMessage
  | WorkerRegisteredMessage

File: packages/python/iii/src/iii/bridge_types.py
"""Bridge message types."""

from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class MessageType(str, Enum):
    """Message types for bridge communication."""

    REGISTER_FUNCTION = "registerfunction"
    REGISTER_SERVICE = "registerservice"
    INVOKE_FUNCTION = "invokefunction"
    INVOCATION_RESULT = "invocationresult"
    REGISTER_TRIGGER_TYPE = "registertriggertype"
    REGISTER_TRIGGER = "registertrigger"
    UNREGISTER_TRIGGER = "unregistertrigger"
    UNREGISTER_TRIGGER_TYPE = "unregistertriggertype"
    TRIGGER_REGISTRATION_RESULT = "triggerregistrationresult"


class RegisterTriggerTypeMessage(BaseModel):
    """Message for registering a trigger type."""

    id: str
    description: str
    type: MessageType = MessageType.REGISTER_TRIGGER_TYPE


class UnregisterTriggerTypeMessage(BaseModel):
    """Message for unregistering a trigger type."""

    id: str
    type: MessageType = MessageType.UNREGISTER_TRIGGER_TYPE


class UnregisterTriggerMessage(BaseModel):
    """Message for unregistering a trigger."""

    id: str
    type: MessageType = MessageType.UNREGISTER_TRIGGER


class TriggerRegistrationResultMessage(BaseModel):
    """Message for trigger registration result."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    trigger_type: str = Field()
    function_id: str = Field()
    result: Any = None
    error: Any = None
    type: MessageType = MessageType.TRIGGER_REGISTRATION_RESULT


class RegisterTriggerMessage(BaseModel):
    """Message for registering a trigger."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    trigger_type: str = Field()
    function_id: str = Field()
    config: Any
    type: MessageType = MessageType.REGISTER_TRIGGER


class RegisterServiceMessage(BaseModel):
    """Message for registering a service."""

    model_config = ConfigDict(populate_by_name=True)

    id: str
    description: str | None = None
    parent_service_id: str | None = Field(default=None)
    type: MessageType = MessageType.REGISTER_SERVICE


class RegisterFunctionFormat(BaseModel):
    """Format definition for function parameters."""

    name: str
    type: str  # 'string' | 'number' | 'boolean' | 'object' | 'array' | 'null' | 'map'
    description: str | None = None
    body: list["RegisterFunctionFormat"] | None = None
    items: "RegisterFunctionFormat | None" = None
    required: bool = False


class RegisterFunctionMessage(BaseModel):
    """Message for registering a function."""

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field()
    description: str | None = None
    request_format: RegisterFunctionFormat | None = Field(default=None)
    response_format: RegisterFunctionFormat | None = Field(default=None)
    metadata: dict[str, Any] | None = None
    type: MessageType = MessageType.REGISTER_FUNCTION


class InvokeFunctionMessage(BaseModel):
    """Message for invoking a function."""

    model_config = ConfigDict(populate_by_name=True)

    function_id: str = Field()
    data: Any
    invocation_id: str | None = Field(default=None)
    type: MessageType = MessageType.INVOKE_FUNCTION


class InvocationResultMessage(BaseModel):
    """Message for invocation result."""

    model_config = ConfigDict(populate_by_name=True)

    invocation_id: str = Field()
    function_id: str = Field()
    result: Any = None
    error: Any = None
    type: MessageType = MessageType.INVOCATION_RESULT


class FunctionInfo(BaseModel):
    """Information about a registered function."""

    function_id: str
    description: str | None = None
    request_format: RegisterFunctionFormat | None = None
    response_format: RegisterFunctionFormat | None = None
    metadata: dict[str, Any] | None = None


WorkerStatus = Literal["connected", "available", "busy", "disconnected"]


class WorkerInfo(BaseModel):
    """Information about a connected worker."""

    id: str
    name: str | None = None
    runtime: str | None = None
    version: str | None = None
    os: str | None = None
    ip_address: str | None = None
    status: WorkerStatus
    connected_at_ms: int
    function_count: int
    functions: list[str]
    active_invocations: int


BridgeMessage = (
    RegisterFunctionMessage
    | InvokeFunctionMessage
    | InvocationResultMessage
    | RegisterServiceMessage
    | RegisterTriggerMessage
    | RegisterTriggerTypeMessage
    | UnregisterTriggerMessage
    | UnregisterTriggerTypeMessage
    | TriggerRegistrationResultMessage
)
